File: src/0230-kth-smallest-element-in-a-bst/solution.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        """
        I wonder if theres a O(k) space solution, as this DFS in-order traversal is just
        O(n) space.

        Note: we use inorder traversal as we want a sorted list to get the kth value
        TO answer the qns, how to modify the BST if theres insert and delete operations:

        we can maintain the size of each subtree in every node, which allows us to efficiently perform order statistics. Then when we add or delete a node, update the node's
        size. Then when we search kth element, we can just do k == node.size
        """
        res = []
        def inorder(node: Optional[TreeNode], k: int):
            if not node:
                return
            inorder(node.left, k)
            res.append(node.val)
            inorder(node.right, k)
        inorder(root, k)
        return res[k - 1]

File: src/0230-kth-smallest-element-in-a-bst/solution.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function kthSmallest(root: TreeNode | null, k: number): number {
    /**
    * Not sure if theres a O(1) space solution. But the O(n) way would be to just sort using
    * in-order traversal: which is left -> middle -> right
    */
    const result = new Array<number>();
    function inOrderTraversal(node: TreeNode | null): void {
        if (!node) {
            return;
        }
        inOrderTraversal(node.left);
        result.push(node.val);
        inOrderTraversal(node.right);
    }
    inOrderTraversal(root);
    console.log(result);
    return result[k - 1];

};
